onnxruntime/core/providers/cuda/nn/instance_norm.cc

```diff
@@ -160,5 +161,150 @@ Status InstanceNorm<T>::ComputeInternal(OpKernelContext* p_op_kernel_context) co
   return Status::OK();
 }
 
+template <>
+Status InstanceNorm<MLFloat16>::ComputeInternal(OpKernelContext* p_op_kernel_context) const {
+  typedef typename ToCudaType<MLFloat16>::MappedType CudaT;
+
+  const Tensor* X = p_op_kernel_context->Input<Tensor>(0);
+  const Tensor* scale = p_op_kernel_context->Input<Tensor>(1);
+  const Tensor* bias = p_op_kernel_context->Input<Tensor>(2);
+
+  ORT_RETURN_IF_ERROR(InstanceNormHelper::ValidateInputs(X, scale, bias));
+
+  const TensorShape& x_shape = X->Shape();
+  Tensor* Y = p_op_kernel_context->Output(0, x_shape);
+
+  auto y_data = reinterpret_cast<CudaT*>(Y->template MutableData<MLFloat16>());
```

Comment:
Same as the other data types in terms of core implementation. Only some auxiliary buffers need to get casted to float32 before calling into CuDNN. This specialization handles that.